src/sync/deque.rs

```diff
@@ -0,0 +1,577 @@
+// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! A lock-free concurrent work-stealing deque
+//!
+//! This module contains an implementation of the Chase-Lev work stealing deque
+//! described in "Dynamic Circular Work-Stealing Deque". The implementation is
+//! heavily based on the pseudocode found in the paper.
+//!
+//! # Example
+//!
+//! ```
+//! use crossbeam::sync::deque;
+//! let (mut worker, stealer) = deque::deque();
+//!
+//! // Only the worker may push/pop
+//! worker.push(1);
+//! worker.pop();
+//!
+//! // Stealers take data from the other end of the deque
+//! worker.push(1);
+//! stealer.steal();
+//!
+//! // Stealers can be cloned to have many stealers stealing in parallel
+//! worker.push(1);
+//! let stealer2 = stealer.clone();
+//! stealer2.steal();
+//! ```
+
+// FIXME: all atomic operations in this module use a SeqCst ordering. That is
+//      probably overkill
+
+use std::cell::UnsafeCell;
+use std::mem;
+use std::ptr;
+use std::sync::atomic::Ordering::SeqCst;
+use std::sync::atomic::AtomicIsize;
+use std::sync::Arc;
+
+use mem::epoch::{self, Atomic, Shared, Owned};
+
+// Once the queue is less than 1/K full, then it will be downsized. Note that
+// the deque requires that this number be less than 2.
+const K: isize = 4;
+
+// Minimum number of bits that a buffer size should be. No buffer will resize to
+// under this value, and all deques will initially contain a buffer of this
+// size.
+//
+// The size in question is 1 << MIN_BITS
+const MIN_BITS: u32 = 7;
+
+struct Deque<T> {
```

Comment:
Currently this isn't public, but in theory it can be so long as the `push` and `pop` methods are `unsafe` (as they only work if one thread is calling them).

For now I figured it could at least start out with the safe API that imposes an `Arc` allocation (the `deque` constructor) and those could perhaps all be deprecated in the future if necessary?